### ISSUE TITLE
Occasionally dropped YouTube titles

### ISSUE THREAD

**Issue by nuxeh** [nuxeh/url-bot-rs#257]
Inexplicably, youtube titles are dropped sometimes, e.g.

```
21:23 <user> https://www.youtube.com/watch?v=YFic-xaLsPs
21:23 <wall-e> ⤷ YouTube
```

**Comment by Samuel-Berger** [comment]
Youtube seems to be the first response when opening a new tab with a youtube URL. Something like a setting for minimum redirect could be in handy, since more steps resolves the tab-title to the title of the video.

**Comment by nuxeh** [comment]
Currently, #319 is available which has a crude hack to do kind of exactly this. If it gets "YouTube" as a title, it retries the request. This was working for a while (but with slow response times due to retry delays), until in the last couple of days it's stopped working...

However, with the weird things going on with getting titles from YouTube recently, the idea has been put forward to start using the YouTube API to get titles instead, which would just require an API key and some JSON requests with any luck, and would actually be more efficient than scraping the title. The plan is to at least give this a go soon.

**Comment by Samuel-Berger** [comment]
I’ll try the workaround. 

I was thinking about using something like https://crates.io/crates/yt_info to extract the information. But I’m new to Rust and don’t have a lot of safe time. So we’ll see how it goes.

Right now I’m in the process of upgrading the dependencies of the crate.

**Comment by nuxeh** [comment]
Thanks for pointing that crate out. Had a little look at it, but it seems fairly unmaintained and seems to be based on `hyper`. This is pretty much the existing idea, but it might be preferable to use the higher level library (based on `hyper`), `reqwest`, and `serde` for JSON decoding. `url-bot-rs` already uses `reqwest`, and so no extra dependencies would be required. So I don't think this crate would be directly useful, it could be a useful reference, however.

**Comment by nuxeh** [comment]
Also, by the way, I'm working on refactoring the HTTP request part of `url-bot-rs` on and off at the moment, so fairly significant changes will probably be going on with that part in in the short term.

**Comment by nuxeh** [comment]
Also I'm working towards making a fairly simple plugin system which could host a YouTube plugin, I have written one working plugin for Imgur galleries already, but the whole thing needs polishing still.

**Comment by nuxeh** [comment]
#330 is now available for using the YouTube API for getting titles.

**Comment by nuxeh** [comment]
@Samuel-Berger I pushed the above PR/branch which has an implementation of YouTube title fetching via the YouTube API, which seems to be working quite well so far, in case you're still interested!

**Comment by nuxeh** [comment]
Youtube plugin (#330) is now merged with `v0.3.1`
